storage/merkle/proof.go

```diff
@@ -0,0 +1,133 @@
+	}
+	// number of steps should be smaller than number of bits in the IsAShortNode
+	if steps > len(p.IsAShortNode)*8 {
+		return false, NewErrorMalformedProoff("the length of IsAShortNode doesn't match the length of SkipBits and InterimHashes")
+	}
+
+	// an index to consume interim hashes from the last element to the first element
+	interimHashIndex := len(p.InterimHashes) - 1
+	// an index to consume shortBits from the last element to the first element
+	skipBitIndex := len(p.SkipBits) - 1
+
+	// keyIndex keeps track of the largest index of the key that is unchecked.
+	// note that traverse bottom up here, so we start with the largest key index
+	// build hashes until we reach to the root.
+	keyIndex := len(p.InterimHashes)
+	for _, sc := range p.SkipBits {
+		keyIndex += int(sc)
+	}
+	keyIndex-- // consider index starts from zero
+
+	if len(p.Key)*8 != keyIndex+1 {
+		return false, NewErrorMalformedProoff("key length doesn't match the length of SkipBits and InterimHashes")
+	}
+
+	// compute the hash value of the leaf
+	currentHash := computeLeafHash(p.Value)
+
+	// for each step (level from bottom to top) check if its a full node or a short node and compute the
+	// hash value accordingly; for full node having the sibling hash helps to compute the hash value
+	// of the next level, for short nodes compute the hash using the common path constructed based on
+	// the given short count
+	for isAShortNodeIndex := steps - 1; isAShortNodeIndex >= 0; isAShortNodeIndex-- {
+
+		// Full node
+		if bitutils.ReadBit(p.IsAShortNode, isAShortNodeIndex) == 0 {
+
+			// read and pop the sibling hash value from InterimHashes
+			if interimHashIndex < 0 {
+				return false, NewErrorMalformedProoff("no more InterimHashes available to read")
+			}
+			sibling := p.InterimHashes[interimHashIndex]
+			interimHashIndex--
+
+			// based on the bit at pathIndex of the key compute the hash
+			if bitutils.ReadBit(p.Key, keyIndex) == 0 { // left branching
+				currentHash = computeFullHash(currentHash, sibling)
+			} else {
+				currentHash = computeFullHash(sibling, currentHash) // right branching
+			}
+
+			// decrement the keyindex by 1
+			keyIndex--
+
+			continue
+		}
+
+		// Short node
+
+		// read and pop from SkipBits
+		if skipBitIndex < 0 {
+			return false, NewErrorMalformedProoff("no more SkipBits available to read")
+		}
+		skipBits := int(p.SkipBits[skipBitIndex])
+		skipBitIndex--
+
+		// construct the common path
+		startIndexOfCommonPath := keyIndex - skipBits + 1
+		commonPath := bitutils.MakeBitVector(skipBits)
+		for j := 0; j < skipBits; j++ {
+			if bitutils.ReadBit(p.Key, startIndexOfCommonPath+j) == 1 {
+				bitutils.SetBit(commonPath, j)
+			}
+		}
+
+		// compute the hash for the short node
+		currentHash = computeShortHash(skipBits, commonPath, currentHash)
+
+		// decrement the keyIndex by the number of bits that were skiped in the short node
+		keyIndex -= skipBits
```

Comment:
Personally, I find it easier to understand, if the code walks consistently backwards over the keys: 

```suggestion
		// construct the common path
		commonPath := bitutils.MakeBitVector(skipBits)
		for c := skipBits - 1; c >= 0; c-- {
			if bitutils.ReadBit(p.Key, keyIndex) == 1 {
				bitutils.SetBit(commonPath, c)
			}
			keyIndex--
		}

		// compute the hash for the short node
		currentHash = computeShortHash(skipBits, commonPath, currentHash)
```

Reply:
updated